design-proposals/persist-vmi-firmware-uuid.md

```diff
@@ -0,0 +1,91 @@
+
+## Motivation
+In the current implementation, the firmware UUID of a Virtual Machine Instance (VMI) is generated dynamically during the creation of the VMI. This means that if the generation logic is changed in the future, the VMI may be assigned a different firmware UUID. This behavior can cause issues for applications that depend on a stable UUID.
+Furthermore, we plan to enhance the firmware UUID generation logic to include the VM's Name and Namespace, ensuring uniqueness across different namespaces. However, changing the generation logic could inadvertently assign new UUIDs to existing VMs upon restart, leading to potential disruptions.
+By persisting the firmware UUID in the VM's status, we can maintain consistency across restarts and future-proof existing VMs against changes in the UUID generation logic.
+
+## Goals
+- Persist Firmware UUID: Store the firmware UUID within the VM's status to ensure it remains consistent across VMI restarts.
+- Maintain Backward Compatibility: Ensure that existing VMIs without a persisted firmware UUID continue to operate without issues.
+- Prepare for Future Enhancements: Allow for future changes to the firmware UUID generation logic without impacting existing VMIs.
+
+## Non Goals
+- Changing Firmware UUID Generation Logic Now: This proposal does not include modifying the current firmware UUID generation to incorporate vm.Name and vm.Namespace.
+- Altering VM Spec: No changes will be made to the VM specification or require user intervention to set or manage the firmware UUID.
+
+## Definition of Users
+End Users: Individuals or organizations running VMs and VMIs on KubeVirt who require consistent firmware UUIDs for their applications.
+
+## User Stories
+As an end-user, I expect my VMI to maintain its identity across restarts.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+We will introduce a new field FirmwareUUID in the status section of the VirtualMachine resource.
+The virt-controller will use this field to persist the firmware UUID (if its empty)
+
+## API Examples
+**Example VM Status**
+
+```yaml
+apiVersion: kubevirt.io/v1alpha3
+kind: VirtualMachine
+metadata:
+  name: mytestvm
+spec:
+  runningStrategy: Halted
+  template:
+    spec:
+      architecture: arm64
+      domain:
+        devices:
+          disks:
+          - name: disk1
+            disk:
+              bus: virtio
+          interfaces:
+          - name: network1
+            bridge: {}
+      networks:
+              - name: network1
+                multus:
+                  networkName: mytestvm/mytestvm
+      volumes:
+           - name: disk1
+             emptyDisk:
+               capacity: 50G
+status:
+  created: true
+  desiredGeneration: 2
+  observedGeneration: 2
+  printableStatus: ErrImagePull
+  runStrategy: Once
+  firmwareUUID: "123e4567-e89b-12d3-a456-426614174000"
```

Comment:
Even if we go with this approach we should list a few alternatives for where the new field should reside. Placing it at the top level isn't necessarily the best choice IMO.

An alternative could be to save the UUID as a VM condition:
https://github.com/kubevirt/kubevirt/blob/ced9fe1b5104daa431bed6b65ea690639fe3b79f/staging/src/kubevirt.io/api/core/v1/types.go#L1659-L1660

Conditions are easier to drop in the future, as they do not change the VM's schema.

Reply:
true